include/albatross/models/sparse_gp.h

```diff
@@ -0,0 +1,246 @@
+      : Base(covariance_function, model_name),
+        inducing_point_strategy(inducing_point_strategy_),
+        independent_group_indexing_function(
+            independent_group_indexing_function_){};
+  SparseGaussianProcessRegression(CovFunc &covariance_function,
+                                  const std::string &model_name)
+      : Base(covariance_function, model_name){};
+
+  template <typename FeatureType>
+  auto _fit_impl(const std::vector<FeatureType> &features,
+                 const MarginalDistribution &targets) const {
+    static_assert(std::is_same<IndexingFunction, LeaveOneOut>::value,
+                  "Only FITC is currently implemented, but by allowing "
+                  "arbitrary groups this could turn into PITC");
+
+    // Determine the set of inducing points, u.
+    auto u = inducing_point_strategy(features);
+
+    Eigen::Index n = targets.mean.size();
+    Eigen::Index m = static_cast<Eigen::Index>(u.size());
+
+    // Create the covariance matrices we need.
+    Eigen::VectorXd K_ff_diag = this->covariance_function_.diagonal(features);
+    if (targets.has_covariance()) {
+      K_ff_diag += targets.covariance.diagonal();
+    }
+    Eigen::MatrixXd K_fu = this->covariance_function_(features, u);
+    Eigen::MatrixXd K_uu = this->covariance_function_(u);
+
+    auto K_uu_llt = K_uu.llt();
+    // P is such that:
+    //     Q_ff = K_fu K_uu^-1 K_uf
+    //          = K_fu L^-T L^-1 K_uf
+    //          = P^T P
+    Eigen::MatrixXd P = K_uu_llt.matrixL().solve(K_fu.transpose());
+
+    // Efficiently compute the diagonal diag[Q_ff].
+    Eigen::VectorXd Q_ff_diag = P.colwise().squaredNorm();
+
+    Eigen::VectorXd A = K_ff_diag - Q_ff_diag;
+
+    if (A.minCoeff() < 1e-6) {
+      // It's possible that the inducing points will perfectly describe
+      // some of the data, in which case we need to add a bit of extra
+      // noise to make sure lambda is invertible.
+      A += 1e-6 * Eigen::VectorXd::Ones(n);
+    }
+    /*
+     *
+     * The end goal here is to produce a vector, v, and matrix, C, such that
+     * for a prediction, f*, we can do,
+     *
+     *     [f*|f=y] ~ N(K_*u * v , K_** - K_*u * C^-1 * K_u*)
+     *
+     *  and it would match the desired prediction described above,
+     *
+     *     [f*|f=y] ~ N(K_*u S K_uf^-1 A^-1 y, K_** − Q_** + K_*u S K_u*)
+     *
+     *  we can find v easily,
+     *
+     *     v = S K_uf^-1 A^-1 y
+     *
+     *  and to get C we need to do some algebra,
+     *
+     *     K_** - K_*u * C * K_u* = K_** - Q_** + K_*u S K_u*
+     *                            = K_** - K_*u (K_uu^-1 - S) K_u*
+     *  which leads to:
+     *     C^-1 = (K_uu^-1 - S)^-1
+     *                                                  (Expansion of S)
+     *          = (K_uu^-1 - (K_uu + K_uf A^-1 K_fu)^-1)^-1
+     *                                        (Woodbury Matrix Identity)
+     *          = (K_uu^-1 K_uf (A + K_fu K_uu^-1 K_uf)^-1 K_fu K_uu^-1)
```

Comment:
I'm pretty sure this is missing a `^-1` at the end, since I think the Woodbury identity gives:

`K_uu^-1 - (K_uu + K_uf A^-1 K_fu )^-1 = K_uu^-1 K_uf ( A + K_fu K_uu^-1 K_uf )^-1 K_fu K_uu^-1`

and we want the inverse of the left hand side.

Reply:
I think this is OK right?  The left hand side is `C^-1`.

Reply:
Yeah, I was talking about the final inverse. I would have assumed that passing the test would mean it's right, but I struggle to see how it can be, considering the code seems to match the comments. Unless `C^-1=C` for some odd reason.